flink-tests/src/test/java/org/apache/flink/test/scheduling/JMFailoverITCase.java

```diff
@@ -0,0 +1,839 @@
+                break;
+            }
+
+            Thread.sleep(100L);
+        }
+
+        // trigger jm failover.
+        triggerJMFailover(jobId);
+
+        // going source 0.
+        SourceTail.goingSubTasks(0);
+
+        JobResult jobResult = flinkCluster.requestJobResult(jobId).get();
+        if (jobResult.getSerializedThrowable().isPresent()) {
+            throw new AssertionError(jobResult.getSerializedThrowable().get());
+        }
+
+        // check count Results
+        checkCountResults();
+    }
+
+    @Test
+    public void testTaskExecutorNotRegisterOnTime() throws Exception {
+        Configuration configuration = new Configuration();
+        configuration.set(
+                BatchExecutionOptions.JOB_RECOVERY_PREVIOUS_WORKER_RECOVERY_TIMEOUT, Duration.ZERO);
+        JobGraph jobGraph = prepareEnvAndGetJobGraph(configuration);
+
+        // blocking all sink
+        StubRecordSink.blockingSubTasks(0, 1, 2, 3);
+
+        JobID jobId = flinkCluster.submitJob(jobGraph).get().getJobID();
+
+        // wait until sink is running.
+        tryWaitUntilCondition(() -> StubRecordSink.attemptIds.size() > 0);
+
+        // trigger jm failover.
+        triggerJMFailover(jobId);
+
+        // going all sink.
+        StubRecordSink.goingSubTasks(0, 1, 2, 3);
+
+        JobResult jobResult = flinkCluster.requestJobResult(jobId).get();
+        if (jobResult.getSerializedThrowable().isPresent()) {
+            throw new AssertionError(jobResult.getSerializedThrowable().get());
+        }
+
+        // check count Results
+        checkCountResults();
+    }
+
+    //    @Ignore
+    @Test
+    public void testPartitionNotFoundTwice() throws Exception {
+        JobGraph jobGraph = prepareEnvAndGetJobGraph();
+
+        // blocking map 0 and map 1.
+        StubMapFunction.blockingSubTasks(0, 1);
+
+        JobID jobId = flinkCluster.submitJob(jobGraph).get().getJobID();
+
+        // wait until map deploying, which indicates all source finished.
+        tryWaitUntilCondition(() -> StubMapFunction.attemptIds.size() > 0);
+
+        triggerJMFailover(jobId);
+
+        // trigger partition not found.
+        releaseResultPartitionOfSource();
+
+        // map 0 going.
+        StubMapFunction.goingSubTasks(0);
+
+        // wait until map 0 restart, which indicates all source finished again.
+        tryWaitUntilCondition(() -> StubMapFunction.attemptIds.get(0) == 1);
+
+        // trigger partition not found.
+        releaseResultPartitionOfSource();
+
+        // map 1 going.
+        StubMapFunction.goingSubTasks(1);
+
+        JobResult jobResult = flinkCluster.requestJobResult(jobId).get();
+        if (jobResult.getSerializedThrowable().isPresent()) {
+            throw new AssertionError(jobResult.getSerializedThrowable().get());
+        }
+
+        // check count Results
+        checkCountResults();
+    }
+
+    @Test
+    public void testPartitionNotFoundAfterJMFailoverWithOperatorCoorUnsupportedBatchSnapshot()
```

Comment:
maybe
`testPartitionNotFoundAndOperatorCoordinatorNotSupportBatchSnapshot`

and `testPartitionNotFoundAfterJMFailoverWithOperatorCoorsupportsBatchSnapshot` -> `testPartitionNotFoundAndOperatorCoordinatorSupportBatchSnapshot`